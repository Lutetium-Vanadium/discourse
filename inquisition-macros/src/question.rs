use std::fmt;

use proc_macro2::TokenStream;
use quote::{quote, quote_spanned};
use syn::{parse::Parse, spanned::Spanned, Token};

use crate::helpers::*;

bitflags::bitflags! {
    pub struct BuilderMethods: u8 {
        const DEFAULT   = 0b000_0001;
        const TRANSFORM = 0b000_0010;
        const VAL_FIL   = 0b000_0100;
        const LIST      = 0b000_1000;
        const MASK      = 0b001_0000;
        const EXTENSION   = 0b010_0000;
        const PLUGIN    = 0b100_0000;
    }
}

#[derive(Clone, Copy)]
pub(crate) enum QuestionKind {
    Input,
    Int,
    Float,
    Confirm,
    List,
    Rawlist,
    Expand,
    Checkbox,
    Password,
    Editor,
    Plugin,
}

impl QuestionKind {
    fn as_str(&self) -> &str {
        match self {
            QuestionKind::Input => "input",
            QuestionKind::Int => "int",
            QuestionKind::Float => "float",
            QuestionKind::Confirm => "confirm",
            QuestionKind::List => "list",
            QuestionKind::Rawlist => "rawlist",
            QuestionKind::Expand => "expand",
            QuestionKind::Checkbox => "checkbox",
            QuestionKind::Password => "password",
            QuestionKind::Editor => "editor",
            QuestionKind::Plugin => "plugin",
        }
    }

    fn get_builder_methods(&self) -> BuilderMethods {
        match *self {
            QuestionKind::Input | QuestionKind::Int | QuestionKind::Float => {
                BuilderMethods::DEFAULT
                    | BuilderMethods::TRANSFORM
                    | BuilderMethods::VAL_FIL
            }
            QuestionKind::Confirm => {
                BuilderMethods::DEFAULT | BuilderMethods::TRANSFORM
            }
            QuestionKind::List | QuestionKind::Rawlist | QuestionKind::Expand => {
                BuilderMethods::DEFAULT
                    | BuilderMethods::TRANSFORM
                    | BuilderMethods::LIST
            }
            QuestionKind::Checkbox => {
                BuilderMethods::DEFAULT
                    | BuilderMethods::TRANSFORM
                    | BuilderMethods::VAL_FIL
                    | BuilderMethods::LIST
            }
            QuestionKind::Password => {
                BuilderMethods::TRANSFORM
                    | BuilderMethods::VAL_FIL
                    | BuilderMethods::MASK
            }
            QuestionKind::Editor => {
                BuilderMethods::DEFAULT
                    | BuilderMethods::TRANSFORM
                    | BuilderMethods::VAL_FIL
                    | BuilderMethods::EXTENSION
            }
            QuestionKind::Plugin => BuilderMethods::PLUGIN,
        }
    }
}

impl Parse for QuestionKind {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let ident = input.parse::<syn::Ident>()?;

        let kind = if ident == "input" {
            QuestionKind::Input
        } else if ident == "int" {
            QuestionKind::Int
        } else if ident == "float" {
            QuestionKind::Float
        } else if ident == "confirm" {
            QuestionKind::Confirm
        } else if ident == "list" {
            QuestionKind::List
        } else if ident == "rawlist" {
            QuestionKind::Rawlist
        } else if ident == "expand" {
            QuestionKind::Expand
        } else if ident == "checkbox" {
            QuestionKind::Checkbox
        } else if ident == "password" {
            QuestionKind::Password
        } else if ident == "editor" {
            QuestionKind::Editor
        } else if ident == "plugin" {
            QuestionKind::Plugin
        } else {
            return Err(syn::Error::new(
                ident.span(),
                format!("unknown question kind {}", ident),
            ));
        };

        Ok(kind)
    }
}

impl fmt::Display for QuestionKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

pub(crate) struct QuestionOpts {
    pub(crate) message: Option<syn::Expr>,
    pub(crate) when: Option<syn::Expr>,
    pub(crate) ask_if_answered: Option<syn::Expr>,

    pub(crate) default: Option<syn::Expr>,

    pub(crate) validate: Option<syn::Expr>,
    pub(crate) filter: Option<syn::Expr>,
    pub(crate) transform: Option<syn::Expr>,
    pub(crate) validate_async: Option<syn::Expr>,
    pub(crate) filter_async: Option<syn::Expr>,
    pub(crate) transform_async: Option<syn::Expr>,

    pub(crate) choices: Option<Choices>,
    pub(crate) page_size: Option<syn::Expr>,
    pub(crate) should_loop: Option<syn::Expr>,

    pub(crate) mask: Option<syn::Expr>,
    pub(crate) extension: Option<syn::Expr>,

    pub(crate) plugin: Option<syn::Expr>,
}

impl Default for QuestionOpts {
    fn default() -> Self {
        Self {
            message: None,
            when: None,
            ask_if_answered: None,

            default: None,

            validate: None,
            filter: None,
            transform: None,
            validate_async: None,
            filter_async: None,
            transform_async: None,

            choices: None,
            page_size: None,
            should_loop: None,

            mask: None,
            extension: None,

            plugin: None,
        }
    }
}

/// Checks if a _valid_ ident is disallowed
fn check_disallowed(
    ident: &syn::Ident,
    kind: QuestionKind,
    allowed: BuilderMethods,
) -> syn::Result<()> {
    #[rustfmt::skip]
    fn disallowed(ident: &syn::Ident, allowed: BuilderMethods) -> bool {
         (ident == "default" &&
          !allowed.contains(BuilderMethods::DEFAULT)) ||

        ((ident == "transform_async" ||
          ident == "transform") &&
          !allowed.contains(BuilderMethods::TRANSFORM)) ||

        ((ident == "validate_async" ||
          ident == "validate" ||
          ident == "filter" ||
          ident == "filter_async") &&
          !allowed.contains(BuilderMethods::VAL_FIL)) ||

        ((ident == "choices" ||
          ident == "page_size" ||
          ident == "should_loop") &&
          !allowed.contains(BuilderMethods::LIST)) ||

         (ident == "mask" &&
          !allowed.contains(BuilderMethods::MASK)) ||

         (ident == "extension" &&
          !allowed.contains(BuilderMethods::EXTENSION)) ||

         (ident == "plugin" &&
          !allowed.contains(BuilderMethods::PLUGIN))
    }

    if disallowed(ident, allowed) {
        Err(syn::Error::new(
            ident.span(),
            format!("option `{}` does not exist for kind `{}`", ident, kind),
        ))
    } else {
        Ok(())
    }
}

pub(crate) struct Question {
    pub(crate) kind: QuestionKind,
    pub(crate) name: syn::Expr,
    pub(crate) opts: QuestionOpts,
}

impl Parse for Question {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let kind = QuestionKind::parse(input)?;
        let content;
        let brace = syn::braced!(content in input);

        let mut opts = QuestionOpts::default();
        let mut name = None;

        let allowed_methods = kind.get_builder_methods();

        while !content.is_empty() {
            if content.peek(Token![async]) {
                let asynct = content.parse::<Token![async]>()?;
                let ident = content.parse::<syn::Ident>()?;

                content.parse::<Token![:]>()?;

                let full_ident_str = format!("{}_async", ident);
                let full_ident = syn::Ident::new(
                    &full_ident_str,
                    asynct
                        .span
                        .join(ident.span())
                        .unwrap_or_else(|| ident.span()),
                );

                check_disallowed(&full_ident, kind, allowed_methods)?;
                if ident == "validate" {
                    insert_non_dup(full_ident, &mut opts.validate_async, &content)?;
                } else if ident == "filter" {
                    insert_non_dup(full_ident, &mut opts.filter_async, &content)?;
                } else if ident == "transform" {
                    insert_non_dup(full_ident, &mut opts.transform_async, &content)?;
                } else {
                    return Err(syn::Error::new(
                        ident.span(),
                        format!("unknown question option `{}`", full_ident_str),
                    ));
                }
            } else {
                let ident = content.parse::<syn::Ident>()?;

                content.parse::<Token![:]>()?;

                // it is not an issue if ident doesn't correspond to valid option
                // since check_allowed only checks if valid idents are disallowed
                check_disallowed(&ident, kind, allowed_methods)?;

                // default options which are always there
                if ident == "name" {
                    insert_non_dup(ident, &mut name, &content)?;
                } else if ident == "message" {
                    insert_non_dup(ident, &mut opts.message, &content)?;
                } else if ident == "when" {
                    insert_non_dup(ident, &mut opts.when, &content)?;
                } else if ident == "ask_if_answered" {
                    insert_non_dup(ident, &mut opts.ask_if_answered, &content)?;
                } else {
                    // the rest may or may not be there, so must be checked
                    // it is not an issue if ident doesn't correspond to valid option
                    // since check_allowed only checks if valid idents are disallowed
                    check_disallowed(&ident, kind, allowed_methods)?;

                    if ident == "default" {
                        insert_non_dup(ident, &mut opts.default, &content)?;
                    } else if ident == "validate" {
                        insert_non_dup(ident, &mut opts.validate, &content)?;
                    } else if ident == "filter" {
                        insert_non_dup(ident, &mut opts.filter, &content)?;
                    } else if ident == "transform" {
                        insert_non_dup(ident, &mut opts.transform, &content)?;
                    } else if ident == "choices" {
                        match kind {
                            QuestionKind::Checkbox => insert_non_dup_parse(
                                ident,
                                &mut opts.choices,
                                &content,
                                Choices::parse_checkbox_choice,
                            )?,
                            _ => insert_non_dup_parse(
                                ident,
                                &mut opts.choices,
                                &content,
                                Choices::parse_choice,
                            )?,
                        }
                    } else if ident == "page_size" {
                        insert_non_dup(ident, &mut opts.page_size, &content)?;
                    } else if ident == "should_loop" {
                        insert_non_dup(ident, &mut opts.should_loop, &content)?;
                    } else if ident == "mask" {
                        insert_non_dup(ident, &mut opts.mask, &content)?;
                    } else if ident == "extension" {
                        insert_non_dup(ident, &mut opts.extension, &content)?;
                    } else if ident == "plugin" {
                        insert_non_dup(ident, &mut opts.plugin, &content)?;
                    } else {
                        return Err(syn::Error::new(
                            ident.span(),
                            format!("unknown question option `{}`", ident),
                        ));
                    }
                }
            }

            if parse_optional_comma(&content)?.is_none() {
                break;
            }
        }

        if let QuestionKind::Plugin = kind {
            if opts.plugin.is_none() {
                return Err(syn::Error::new(
                    brace.span,
                    "missing required option `plugin`",
                ));
            }
        }

        Ok(Self {
            kind,
            name: name.ok_or_else(|| {
                syn::Error::new(brace.span, "missing required option `name`")
            })?,
            opts,
        })
    }
}

impl Question {
    fn write_main_opts(&self, tokens: &mut TokenStream) {
        if let Some(ref message) = self.opts.message {
            tokens.extend(quote_spanned! { message.span() => .message(#message) });
        }
        if let Some(ref when) = self.opts.when {
            tokens.extend(quote_spanned! { when.span() => .when(#when) });
        }
        if let Some(ref ask_if_answered) = self.opts.ask_if_answered {
            tokens.extend(quote_spanned! {
                ask_if_answered.span() => .ask_if_answered(#ask_if_answered)
            });
        }
    }
}

impl quote::ToTokens for Question {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let name = &self.name;

        if let QuestionKind::Plugin = self.kind {
            let plugin = self.opts.plugin.as_ref().unwrap();
            // If just the name was passed into Question::plugin, type errors associated
            // with its conversion to a string would take the span _including_ that of
            // plugin. Explicitly performing `String::from`, makes the error span due to
            // the `From` trait will show the span of the name only
            let name = quote_spanned! {
                name.span() => String::from(#name)
            };
            tokens.extend(quote_spanned! {
                plugin.span() => ::inquisition::Question::plugin(#name, #plugin)
            });
            self.write_main_opts(tokens);
            tokens.extend(quote! { .build() });
            return;
        }

        let kind = syn::Ident::new(self.kind.as_str(), name.span());

        tokens.extend(quote_spanned! {
            name.span() => ::inquisition::Question::#kind(#name)
        });

        self.write_main_opts(tokens);
        if let Some(ref default) = self.opts.default {
            tokens.extend(quote_spanned! { default.span() => .default(#default) });
        }
        if let Some(ref validate) = self.opts.validate {
            tokens
                .extend(quote_spanned! { validate.span() => .validate(#validate) });
        }
        if let Some(ref validate_async) = self.opts.validate_async {
            tokens.extend(quote_spanned! {
                validate_async.span() => .validate_async(#validate_async)
            });
        }
        if let Some(ref filter) = self.opts.filter {
            tokens.extend(quote_spanned! { filter.span() => .filter(#filter) });
        }
        if let Some(ref filter_async) = self.opts.filter_async {
            tokens.extend(quote_spanned! {
                filter_async.span() => .filter_async(#filter_async)
            });
        }
        if let Some(ref transform) = self.opts.transform {
            tokens.extend(
                quote_spanned! { transform.span() => .transform(#transform) },
            );
        }
        if let Some(ref transform_async) = self.opts.transform_async {
            tokens.extend(quote_spanned! {
                transform_async.span() => .transform_async(#transform_async)
            });
        }
        if let Some(ref choices) = self.opts.choices {
            tokens.extend(match self.kind {
                QuestionKind::Checkbox => {
                    quote_spanned! { choices.span() => .choices_with_default(#choices) }
                }
                _ => quote_spanned! { choices.span() => .choices(#choices) },
            });
        }
        if let Some(ref page_size) = self.opts.page_size {
            tokens.extend(
                quote_spanned! { page_size.span() => .page_size(#page_size) },
            );
        }
        if let Some(ref should_loop) = self.opts.should_loop {
            tokens.extend(
                quote_spanned! { should_loop.span() => .should_loop(#should_loop) },
            );
        }
        if let Some(ref mask) = self.opts.mask {
            tokens.extend(quote_spanned! { mask.span() => .mask(#mask) });
        }
        if let Some(ref extension) = self.opts.extension {
            tokens.extend(
                quote_spanned! { extension.span() => .extension(#extension) },
            );
        }
        tokens.extend(quote! { .build() });
    }
}
