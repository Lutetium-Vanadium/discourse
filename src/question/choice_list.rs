use std::ops::{Index, IndexMut};

pub(crate) struct ChoiceList<T> {
    pub(crate) choices: Vec<Choice<T>>,
    pub(crate) default: usize,
    pub(crate) should_loop: bool,
    pub(crate) page_size: usize,
}

impl<T> ChoiceList<T> {
    pub(crate) fn len(&self) -> usize {
        self.choices.len()
    }
}

impl<T> Index<usize> for ChoiceList<T> {
    type Output = Choice<T>;

    fn index(&self, index: usize) -> &Self::Output {
        &self.choices[index]
    }
}

impl<T> IndexMut<usize> for ChoiceList<T> {
    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
        &mut self.choices[index]
    }
}

pub enum Choice<T> {
    Choice(T),
    Separator(Option<String>),
}

impl<T> Choice<T> {
    pub(crate) fn is_separator(&self) -> bool {
        matches!(self, Choice::Separator(_))
    }

    pub(crate) fn as_ref(&self) -> Choice<&T> {
        match self {
            Choice::Choice(t) => Choice::Choice(t),
            Choice::Separator(s) => Choice::Separator(s.clone()),
        }
    }

    pub(crate) fn unwrap_choice(self) -> T {
        if let Choice::Choice(c) = self {
            c
        } else {
            panic!("Called unwrap_choice on separator")
        }
    }
}

pub(crate) fn get_sep_str(separator: &Option<String>) -> &str {
    separator
        .as_ref()
        .map(String::as_str)
        .unwrap_or("──────────────")
}

impl<T: AsRef<str>> Choice<T> {
    pub(crate) fn as_str(&self) -> &str {
        match self {
            Choice::Choice(t) => t.as_ref(),
            Choice::Separator(s) => get_sep_str(s),
        }
    }

    pub(crate) fn as_bytes(&self) -> &[u8] {
        self.as_str().as_bytes()
    }
}

impl<T> From<T> for Choice<T> {
    fn from(t: T) -> Self {
        Choice::Choice(t)
    }
}

impl From<&'_ str> for Choice<String> {
    fn from(s: &str) -> Self {
        Choice::Choice(s.into())
    }
}
