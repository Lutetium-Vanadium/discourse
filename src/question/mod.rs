mod checkbox;
mod choice;
mod confirm;
mod editor;
mod expand;
mod input;
mod number;
mod select;
#[macro_use]
mod options;
mod password;
mod plugin;
mod raw_select;

pub use checkbox::CheckboxBuilder;
pub use confirm::ConfirmBuilder;
pub use editor::EditorBuilder;
pub use expand::ExpandBuilder;
pub use input::InputBuilder;
pub use number::{FloatBuilder, IntBuilder};
pub use password::PasswordBuilder;
pub use plugin::PluginBuilder;
pub use raw_select::RawSelectBuilder;
pub use select::SelectBuilder;

use crate::{Answer, Answers};
pub use choice::Choice;
use choice::{get_sep_str, ChoiceList};
use options::Options;
pub use plugin::Plugin;
use plugin::PluginInteral;
use ui::{backend::Backend, error, events::KeyEvent};

use std::fmt;

#[derive(Debug)]
pub struct Question<'a> {
    kind: QuestionKind<'a>,
    opts: Options<'a>,
}

impl<'a> Question<'a> {
    fn new(opts: Options<'a>, kind: QuestionKind<'a>) -> Self {
        Self { kind, opts }
    }
}

impl Question<'static> {
    pub fn input<N: Into<String>>(name: N) -> InputBuilder<'static> {
        InputBuilder::new(name.into())
    }

    pub fn int<N: Into<String>>(name: N) -> IntBuilder<'static> {
        IntBuilder::new(name.into())
    }

    pub fn float<N: Into<String>>(name: N) -> FloatBuilder<'static> {
        FloatBuilder::new(name.into())
    }

    pub fn confirm<N: Into<String>>(name: N) -> ConfirmBuilder<'static> {
        ConfirmBuilder::new(name.into())
    }

    pub fn select<N: Into<String>>(name: N) -> SelectBuilder<'static> {
        SelectBuilder::new(name.into())
    }

    pub fn raw_select<N: Into<String>>(name: N) -> RawSelectBuilder<'static> {
        RawSelectBuilder::new(name.into())
    }

    pub fn expand<N: Into<String>>(name: N) -> ExpandBuilder<'static> {
        ExpandBuilder::new(name.into())
    }

    pub fn checkbox<N: Into<String>>(name: N) -> CheckboxBuilder<'static> {
        CheckboxBuilder::new(name.into())
    }

    pub fn password<N: Into<String>>(name: N) -> PasswordBuilder<'static> {
        PasswordBuilder::new(name.into())
    }

    pub fn editor<N: Into<String>>(name: N) -> EditorBuilder<'static> {
        EditorBuilder::new(name.into())
    }

    pub fn plugin<'a, N, P>(name: N, plugin: P) -> PluginBuilder<'a>
    where
        N: Into<String>,
        P: Plugin + 'a,
    {
        PluginBuilder::new(name.into(), Box::new(Some(plugin)))
    }
}

#[derive(Debug)]
enum QuestionKind<'a> {
    Input(input::Input<'a>),
    Int(number::Int<'a>),
    Float(number::Float<'a>),
    Confirm(confirm::Confirm<'a>),
    Select(select::Select<'a>),
    RawSelect(raw_select::RawSelect<'a>),
    Expand(expand::Expand<'a>),
    Checkbox(checkbox::Checkbox<'a>),
    Password(password::Password<'a>),
    Editor(editor::Editor<'a>),
    Plugin(Box<dyn PluginInteral + 'a>),
}

impl Question<'_> {
    pub(crate) fn ask<B: Backend, I: Iterator<Item = error::Result<KeyEvent>>>(
        mut self,
        answers: &Answers,
        b: &mut B,
        events: &mut I,
    ) -> error::Result<Option<(String, Answer)>> {
        if (!self.opts.ask_if_answered && answers.contains_key(&self.opts.name))
            || !self.opts.when.get(answers)
        {
            return Ok(None);
        }

        let name = self.opts.name;
        let message = self
            .opts
            .message
            .map(|message| message.get(answers))
            .unwrap_or_else(|| name.clone() + ":");

        let res = match self.kind {
            QuestionKind::Input(i) => i.ask(message, answers, b, events)?,
            QuestionKind::Int(i) => i.ask(message, answers, b, events)?,
            QuestionKind::Float(f) => f.ask(message, answers, b, events)?,
            QuestionKind::Confirm(c) => c.ask(message, answers, b, events)?,
            QuestionKind::Select(l) => l.ask(message, answers, b, events)?,
            QuestionKind::RawSelect(r) => r.ask(message, answers, b, events)?,
            QuestionKind::Expand(e) => e.ask(message, answers, b, events)?,
            QuestionKind::Checkbox(c) => c.ask(message, answers, b, events)?,
            QuestionKind::Password(p) => p.ask(message, answers, b, events)?,
            QuestionKind::Editor(e) => e.ask(message, answers, b, events)?,
            QuestionKind::Plugin(ref mut o) => o.ask(message, answers, b, events)?,
        };

        Ok(Some((name, res)))
    }
}

macro_rules! handler {
    ($name:ident, $fn_trait:ident ( $($type:ty),* ) -> $return:ty) => {
        pub(crate) enum $name<'a, T> {
            Sync(Box<dyn $fn_trait( $($type),* ) -> $return + 'a>),
            None,
        }

        impl<'a, T> $name<'a, T> {
            #[allow(unused)]
            fn take(&mut self) -> Self {
                std::mem::replace(self, Self::None)
            }
        }

        impl<T> Default for $name<'_, T> {
            fn default() -> Self {
                Self::None
            }
        }

        impl<T: fmt::Debug> fmt::Debug for $name<'_, T> {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                match self {
                    Self::Sync(_) => f.write_str("Sync(_)"),
                    Self::None => f.write_str("None"),
                }
            }
        }
    };

    // The type signature of the function must only contain &T
    ($name:ident, ?Sized $fn_trait:ident ( $($type:ty),* ) -> $return:ty) => {
        pub(crate) enum $name<'a, T: ?Sized> {
            Sync(Box<dyn $fn_trait( $($type),* ) -> $return + 'a>),
            None,
        }

        impl<'a, T: ?Sized> $name<'a, T> {
            #[allow(unused)]
            fn take(&mut self) -> Self {
                std::mem::replace(self, Self::None)
            }
        }

        impl<T: ?Sized> Default for $name<'_, T> {
            fn default() -> Self {
                Self::None
            }
        }

        impl<T: fmt::Debug + ?Sized> fmt::Debug for $name<'_, T> {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                match self {
                    Self::Sync(_) => f.write_str("Sync(_)"),
                    Self::None => f.write_str("None"),
                }
            }
        }
    };
}

#[cfg(feature = "smallvec")]
pub type Completions<T> = smallvec::SmallVec<[T; 1]>;
#[cfg(not(feature = "smallvec"))]
pub type Completions<T> = Vec<T>;

handler!(Filter, FnOnce(T, &Answers) -> T);
handler!(AutoComplete, FnMut(T, &Answers) -> Completions<T>);
handler!(Validate, ?Sized FnMut(&T, &Answers) -> Result<(), String>);
handler!(ValidateByVal, FnMut(T, &Answers) -> Result<(), String>);
handler!(Transform, ?Sized FnOnce(&T, &Answers, &mut dyn Backend) -> error::Result<()>);
handler!(
    TransformByVal,
    FnOnce(T, &Answers, &mut dyn Backend) -> error::Result<()>
);

#[doc(hidden)]
#[macro_export]
macro_rules! impl_filter_builder {
    ($t:ty; $inner:ident) => {
        pub fn filter<F>(mut self, filter: F) -> Self
        where
            F: FnOnce($t, &crate::Answers) -> $t + 'a,
        {
            self.$inner.filter = crate::question::Filter::Sync(Box::new(filter));
            self
        }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! impl_auto_complete_builder {
    ($t:ty; $inner:ident) => {
        pub fn auto_complete<F>(mut self, auto_complete: F) -> Self
        where
            F: FnMut($t, &crate::Answers) -> Completions<$t> + 'a,
        {
            self.$inner.auto_complete =
                crate::question::AutoComplete::Sync(Box::new(auto_complete));
            self
        }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! impl_validate_builder {
    ($t:ty; $inner:ident) => {
        pub fn validate<F>(mut self, filter: F) -> Self
        where
            F: FnMut(&$t, &crate::Answers) -> Result<(), String> + 'a,
        {
            self.$inner.validate = crate::question::Validate::Sync(Box::new(filter));
            self
        }
    };

    (by val $t:ty; $inner:ident) => {
        pub fn validate<F>(mut self, filter: F) -> Self
        where
            F: FnMut($t, &crate::Answers) -> Result<(), String> + 'a,
        {
            self.$inner.validate = crate::question::ValidateByVal::Sync(Box::new(filter));
            self
        }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! impl_transform_builder {
    ($t:ty; $inner:ident) => {
        pub fn transform<F>(mut self, transform: F) -> Self
        where
            F: FnOnce(&$t, &crate::Answers, &mut dyn Backend) -> ui::error::Result<()> + 'a,
        {
            self.$inner.transform = crate::question::Transform::Sync(Box::new(transform));
            self
        }
    };

    (by val $t:ty; $inner:ident) => {
        pub fn transform<F>(mut self, transform: F) -> Self
        where
            F: FnOnce($t, &crate::Answers, &mut dyn Backend) -> ui::error::Result<()> + 'a,
        {
            self.$inner.transform = crate::question::TransformByVal::Sync(Box::new(transform));
            self
        }
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! write_final {
    ($transform:expr, $message:expr, $ans:expr, $answers:expr, $backend:expr, $custom:expr) => {{
        ui::widgets::Prompt::write_finished_message(&$message, $backend)?;

        match $transform {
            Transform::Sync(transform) => transform($ans, $answers, $backend)?,
            _ => $custom,
        }

        $backend.write_all(b"\n")?;
        $backend.flush()?;
    }};
}
