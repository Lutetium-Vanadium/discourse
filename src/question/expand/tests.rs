use ui::{backend::TestBackend, events::KeyCode, layout::Layout};

use super::*;
use crate::question::QuestionKind;

#[test]
#[should_panic(expected = "Reserved key 'h'")]
fn test_panic_reserved_key() {
    ExpandBuilder::new("name".into()).choice('h', "help");
}

#[test]
#[should_panic(expected = "Duplicate key 'k'")]
fn test_panic_duplicate() {
    ExpandBuilder::new("name".into())
        .choice('k', "key 1")
        .choice('k', "key 2");
}

#[test]
#[should_panic(expected = "Duplicate key 'k'")]
fn test_panic_duplicate_case_insensitive() {
    ExpandBuilder::new("name".into())
        .choice('k', "key 1")
        .choice('K', "key 2");
}

#[test]
#[should_panic(expected = "Invalid default 'd' does not occur in the given choices")]
fn test_panic_invalid_default() {
    ExpandBuilder::new("name".into())
        .choice('k', "key 1")
        .default('d')
        .build();
}

macro_rules! expand {
    (let mut $expand:ident; $message:expr) => {
        expand!(let mut $expand; $message, 'h');
    };

    (let mut $expand:ident; $message:expr, $default:expr) => {
        let hint: String = "abcdefgh"
            .chars()
            .map(|c| {
                if c == $default {
                    c.to_ascii_uppercase()
                } else {
                    c
                }
            })
            .collect();

        let expand =
            match ExpandBuilder::new("name".into())
                .choices(('a'..='g').map(|key| {
                    (key, format!("Choice {}", key.to_ascii_uppercase()))
                }))
                .build()
                .kind
            {
                QuestionKind::Expand(e) => e,
                _ => unreachable!(),
            };

        let mut $expand = ExpandPrompt {
            prompt: widgets::Prompt::new($message).with_hint(&hint),
            input: widgets::CharInput::with_filter_map(|c| {
                let c = c.to_ascii_lowercase();
                hint.chars()
                    .find(|o| o.eq_ignore_ascii_case(&c))
                    .and(Some(c))
            }),
            select: widgets::Select::new(expand),
            expanded: false,
        };
    };
}

#[test]
fn test_render() {
    let size = (50, 20).into();
    let base_layout = Layout::new(0, size);
    let mut backend = TestBackend::new_with_layout(size, base_layout);

    expand!(let mut expand; "message");

    let keys = [
        (
            KeyEvent::from(KeyCode::Char('a')),
            base_layout.with_offset(0, 2),
        ),
        (
            KeyEvent::from(KeyCode::Char('h')),
            base_layout.with_offset(0, 2),
        ),
        (
            KeyEvent::from(KeyCode::Backspace),
            base_layout.with_line_offset(21),
        ),
    ];

    let mut layout = base_layout;
    assert!(expand.render(&mut layout, &mut backend).is_ok());
    ui::assert_backend_snapshot!(backend);
    assert_eq!(layout, base_layout.with_line_offset(21));

    for &(key, expected_layout) in keys.iter() {
        layout = base_layout;
        backend.reset_with_layout(layout);
        assert!(expand.handle_key(key));
        assert!(expand.render(&mut layout, &mut backend).is_ok());
        ui::assert_backend_snapshot!(backend);
        assert_eq!(layout, expected_layout);
    }

    layout = base_layout;
    backend.reset_with_layout(layout);
    assert_eq!(expand.validate(), Ok(Validation::Continue));
    assert!(expand.render(&mut layout, &mut backend).is_ok());
    ui::assert_backend_snapshot!(backend);
    assert_eq!(layout, base_layout.with_offset(0, 9).with_line_offset(10));

    layout = base_layout;
    backend.reset_with_layout(layout);
    assert!(expand.handle_key(KeyCode::Char('c').into()));
    assert!(expand.render(&mut layout, &mut backend).is_ok());
    ui::assert_backend_snapshot!(backend);
    assert_eq!(layout, base_layout.with_offset(0, 9).with_line_offset(11));
}

#[test]
fn test_height() {
    let size = (50, 20).into();
    let base_layout = Layout::new(0, size);

    expand!(let mut expand; "message");

    let keys = [
        (
            KeyEvent::from(KeyCode::Char('a')),
            2,
            base_layout.with_offset(0, 2),
        ),
        (
            KeyEvent::from(KeyCode::Char('h')),
            2,
            base_layout.with_offset(0, 2),
        ),
        (
            KeyEvent::from(KeyCode::Backspace),
            1,
            base_layout.with_line_offset(21),
        ),
    ];

    let mut layout = base_layout;
    assert_eq!(expand.height(&mut layout), 1);
    assert_eq!(layout, base_layout.with_line_offset(21));

    for &(key, height, expected_layout) in keys.iter() {
        layout = base_layout;
        assert!(expand.handle_key(key));
        assert_eq!(expand.height(&mut layout), height);
        assert_eq!(layout, expected_layout);
    }

    layout = base_layout;
    assert_eq!(expand.validate(), Ok(Validation::Continue));
    assert_eq!(expand.height(&mut layout), 10);
    assert_eq!(layout, base_layout.with_offset(0, 9).with_line_offset(10));

    layout = base_layout;
    assert!(expand.handle_key(KeyCode::Char('c').into()));
    assert_eq!(expand.height(&mut layout), 10);
    assert_eq!(layout, base_layout.with_offset(0, 9).with_line_offset(11));
}

#[test]
fn test_cursor_pos() {
    let size = (50, 20).into();
    let layout = Layout::new(5, size);

    expand!(let mut expand; "message");

    let keys = [
        (KeyEvent::from(KeyCode::Char('a')), (27, 0)),
        (KeyEvent::from(KeyCode::Char('h')), (27, 0)),
        (KeyEvent::from(KeyCode::Backspace), (26, 0)),
    ];

    assert_eq!(expand.cursor_pos(layout), (26, 0));

    for &(key, cursor_pos) in keys.iter() {
        assert!(expand.handle_key(key));
        assert_eq!(expand.cursor_pos(layout), cursor_pos);
    }

    assert_eq!(expand.validate(), Ok(Validation::Continue));
    assert_eq!(expand.cursor_pos(layout), (10, 9));

    assert!(expand.handle_key(KeyCode::Char('c').into()));
    assert_eq!(expand.cursor_pos(layout), (11, 9));

    let message = "-".repeat(size.width as usize) + "message";
    expand!(let mut expand; &*message);

    let keys = [
        (KeyEvent::from(KeyCode::Char('a')), (27, 1)),
        (KeyEvent::from(KeyCode::Char('h')), (27, 1)),
        (KeyEvent::from(KeyCode::Backspace), (26, 1)),
    ];

    assert_eq!(expand.cursor_pos(layout), (26, 1));

    for &(key, cursor_pos) in keys.iter() {
        assert!(expand.handle_key(key));
        assert_eq!(expand.cursor_pos(layout), cursor_pos);
    }

    assert_eq!(expand.validate(), Ok(Validation::Continue));
    assert_eq!(expand.cursor_pos(layout), (10, 10));

    assert!(expand.handle_key(KeyCode::Char('c').into()));
    assert_eq!(expand.cursor_pos(layout), (11, 10));
}
