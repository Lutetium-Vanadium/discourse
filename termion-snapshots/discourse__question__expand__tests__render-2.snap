---
source: src/question/expand/tests.rs
expression: backend

---
┌──────────────────────────────────────────────────┐
│[38;5;10m? [1m[39mmessage[22m [38;5;8m(abcdefgH)[39m a                            │
│[38;5;6m>> [39mChoice A                                       │
│[48;5;7m [49m                                                 │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
│                                                  │
└──────────────────────────────────────────────────┘
